irctest/server_tests/redact.py

```diff
@@ -0,0 +1,672 @@
+        alice = random_name("alice")
+        channel = random_name("#channel")
+
+        self.connectClient(
+            alice,
+            name=alice,
+            capabilities=[
+                "message-tags",
+                "echo-message",
+                "batch",
+                "labeled-response",
+                REDACT_CAP,
+            ],
+            skip_if_cap_nak=True,
+        )
+        self.joinChannel(alice, channel)
+        self.getMessages(alice)
+
+        # Send a message
+        self.sendLine(alice, f"PRIVMSG {channel} :Oops wrong channel")
+        echo = self.getMessage(alice)
+        msgid = echo.tags.get("msgid")
+        assert msgid, "Server did not send a msgid tag"
+
+        # Redact with a reason
+        self.sendLine(alice, f"REDACT {channel} {msgid} :wrong channel")
+        redact = self.getMessage(alice)
+
+        self.assertMessageMatch(
+            redact,
+            command="REDACT",
+            params=[channel, msgid, "wrong channel"],
+            prefix=StrRe(f"{alice}!.*"),
+        )
+
+    @pytest.mark.arbitrary_client_tags
+    @cases.mark_capabilities(
+        "message-tags", "echo-message", "batch", "labeled-response", REDACT_CAP
+    )
+    def testRedactTagmsg(self):
+        """Test redaction of a TAGMSG."""
+        alice = random_name("alice")
+        bob = random_name("bob")
+        channel = random_name("#channel")
+
+        # Connect clients with redaction capability
+        self.connectClient(
+            alice,
+            name=alice,
+            capabilities=[
+                "message-tags",
+                "echo-message",
+                "batch",
+                "labeled-response",
+                REDACT_CAP,
+            ],
+            skip_if_cap_nak=True,
+        )
+        self.joinChannel(alice, channel)
+        self.getMessages(alice)
+
+        self.connectClient(
+            bob,
+            name=bob,
+            capabilities=[
+                "message-tags",
+                "echo-message",
+                "batch",
+                "labeled-response",
+                REDACT_CAP,
+            ],
+        )
+        self.joinChannel(bob, channel)
+        self.getMessages(alice)
+        self.getMessages(bob)
+
+        # Send a TAGMSG (e.g., a reaction)
+        self.sendLine(alice, f"@+draft/react=👍 TAGMSG {channel}")
+        echo = self.getMessage(alice)
+        delivery = self.getMessage(bob)
+
+        self.assertMessageMatch(echo, command="TAGMSG", params=[channel])
+        self.assertMessageMatch(delivery, command="TAGMSG", params=[channel])
+
+        msgid = echo.tags.get("msgid")
+        assert msgid, "Server did not send a msgid tag"
+
+        # Redact the TAGMSG
+        self.sendLine(alice, f"REDACT {channel} {msgid}")
+        alice_redact = self.getMessage(alice)
+
+        # Some servers may not support redacting TAGMSGs (e.g., they don't store them)
+        # and return UNKNOWN_MSGID. This is acceptable behavior.
```

Comment:
I wonder if we should hardcode the list of servers that don't support that. It would allow detecting regressions in the others.